book_source/06_reference/04_docker/03_architecture.Rmd

```diff
@@ -17,5 +19,186 @@ The PEcAn framework containers consist of multiple unique ways to interact with
 
 The model containers contain the actual models that are executed as well as small wrappers to make them work in the PEcAn framework. The containers will run the model based on the parameters received from the message bus and convert the outputs back to the standard PEcAn output format. Once the container is finished processing a message it will immediatly get the next message and start processing it.
 
+### PEcAn's `docker-compose` {#pecan-docker-compose}
+
+```{r comment='', echo = FALSE, results = 'hide'}
+docker_compose_file <- file.path("..", "docker-compose.yml")
+dc_yaml <- yaml::read_yaml(docker_compose_file)
+```
+
+The PEcAn Docker architecture is described in full by the PEcAn `docker-compose.yml` file.
+For full `docker-compose` syntax, see the [official documentation](https://docs.docker.com/compose/).
+
+This section describes the [top-level structure](#pecan-dc-structure) and each of the services, which are as follows:
+
+- [`traefik`](#pecan-dc-traefik)
+- [`minio`](#pecan-dc-minio)
+- [`postgres`](#pecan-dc-postgres)
+- [`rabbitmq`](#pecan-dc-rabbitmq)
+- [`bety`](#pecan-dc-bety)
+- [`web`](#pecan-dc-web)
+- [`executor`](#pecan-dc-executor)
+- [Model-specific services](#pecan-dc-models)
+
+For reference, the complete `docker-compose` file is as follows:
+
+```{r, echo = FALSE, comment = ''}
+yaml::write_yaml(dc_yaml, stdout())
+```
+
+#### Top-level structure {#pecan-dc-structure}
+
+The root of the `docker-compose.yml` file contains three sections:
+
+- `services` -- This is a list of services provided by the application, with each service corresponding to a container.
+When communicating with each other internally, the hostnames of containers correspond to their names in this section.
+For instance, regardless of the "project" name passed to `docker-compose up`, the hostname for connecting to the PostgreSQL database of any given container is _always_ going to be `postgres` (e.g. you should be able to access the PostgreSQL database by calling the following from inside the container: `psql -d bety -U bety -h postgres`).
+The services comprising the PEcAn application are described below.
+
+- `networks` -- This is a list of networks used by the application.
```

Comment:
Only if two containers are connected to the same network can they communicate with each other using ports and hostnames. If two containers on different networks they will not be able to talk to each other directly, only using  ports exposed to hosting docker machine.